prow/flagutil/github.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flagutil
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"net/url"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/git"
+	"k8s.io/test-infra/prow/github"
+)
+
+// GitHubOptions holds options for interacting with GitHub.
+type GitHubOptions struct {
+	endpoint            Strings
+	TokenPath           string
+	deprecatedTokenFile string
+}
+
+// AddFlags injects GitHub options into the given FlagSet.
+func (o *GitHubOptions) AddFlags(fs *flag.FlagSet) {
+	o.endpoint = NewStrings("https://api.github.com")
+	fs.Var(&o.endpoint, "github-endpoint", "GitHub's API endpoint (may differ for enterprise).")
+	fs.StringVar(&o.TokenPath, "github-token-path", "/etc/github/oauth", "Path to the file containing the GitHub OAuth secret.")
+	fs.StringVar(&o.deprecatedTokenFile, "github-token-file", "", "DEPRECATED: use -github-token-path instead.  -github-token-file may be removed anytime after 2019-01-01.")
+}
+
+// Validate validates GitHub options.
+func (o *GitHubOptions) Validate(dryRun bool) error {
+	for _, uri := range o.endpoint.Strings() {
+		if _, err := url.ParseRequestURI(uri); err != nil {
+			return fmt.Errorf("invalid -github-endpoint URI: %q", uri)
+		}
+	}
+
+	if o.deprecatedTokenFile != "" {
+		o.TokenPath = o.deprecatedTokenFile
+		logrus.Error("-github-token-file is deprecated and may be removed anytime after 2019-01-01.  Use -github-token-path instead.")
+	}
+
+	if o.TokenPath == "" {
+		return errors.New("empty -github-token-path")
+	}
+
+	return nil
+}
+
+// GitHubClient returns a GitHub client.
+func (o *GitHubOptions) GitHubClient(secretAgent *config.SecretAgent, dryRun bool) (client *github.Client, err error) {
+	var generator *func() []byte
+	if o.TokenPath == "" {
+		generatorFunc := func() []byte {
+			return []byte{}
+		}
+		generator = &generatorFunc
```

Comment:
@wking here if `o.TokenPath` is empty isn't we want to return a nil github client?

Reply:
> @wking here if `o.TokenPath` is empty isn't we want to return a nil github client?

You could still have an unauthenticated client (e.g. if you only needed to make a few requests and weren't worried about rate-limiting).  If that's not a use-case we want to support, I'm ok erroring out here.  But I don't see a point to returning `nil, nil` here.

Reply:
(we have a plank that is not interfacing with github at all :-) so probably we want to keep the old behavior? )

Reply:
> ...we have a plank that is not interfacing with github at all...

So returning an anonymous client (like #9913), sounds reasonable, doesn't it?

Reply:
what's the behavior when the anonymous client making github calls? Errors? :-)